bindings/nodejs/src/sol-uv-integration.c

```diff
@@ -0,0 +1,212 @@
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include <uv.h>
+#include <sol-mainloop.h>
+#include <sol-log.h>
+
+#define RESOLVE_MAINLOOP_STATE(state) \
+    ( state == MAINLOOP_HIJACKING_STARTED ? "MAINLOOP_HIJACKING_STARTED" : \
+    state == MAINLOOP_HIJACKED ? "MAINLOOP_HIJACKED" : \
+    state == MAINLOOP_RELEASING_STARTED ? "MAINLOOP_RELEASING_STARTED" : \
+    state == MAINLOOP_RELEASED ? "MAINLOOP_RELEASED" : "Unknown" )
+
+enum MainloopState {
+    MAINLOOP_HIJACKING_STARTED,
+    MAINLOOP_HIJACKED,
+    MAINLOOP_RELEASING_STARTED,
+    MAINLOOP_RELEASED
+};
+
+static enum MainloopState mainloopState = MAINLOOP_RELEASED;
+static uv_idle_t uv_idle;
+static struct sol_mainloop_source *uv_loop_source = NULL;
+static struct sol_fd *uv_loop_fd = NULL;
+
+static void
+uv_idle_callback()
+{
+    SOL_DBG("Entering with state %s", RESOLVE_MAINLOOP_STATE(mainloopState));
+    if (mainloopState == MAINLOOP_HIJACKING_STARTED) {
+        SOL_DBG("running sol_run()");
+        mainloopState = MAINLOOP_HIJACKED;
+        sol_run();
+        SOL_DBG("sol_run() has returned. state is %s",
+            RESOLVE_MAINLOOP_STATE(mainloopState));
+        if (mainloopState == MAINLOOP_RELEASING_STARTED) {
+            mainloopState = MAINLOOP_RELEASED;
+        }
+    } else if ( mainloopState == MAINLOOP_HIJACKED) {
+        SOL_DBG("main loop already hijacked. Stopping idler");
+        uv_idle_stop(&uv_idle);
+    }
+}
+
+static bool
+uv_loop_source_check(void *data)
+{
+    uv_loop_t *loop = data;
+
+    uv_update_time(loop);
+
+    bool returnValue = uv_loop_alive(loop);
+    SOL_DBG("Returning %s", returnValue ? "true" : "false");
+    return returnValue;
+}
+
+static bool
+uv_loop_source_get_next_timeout(void *data,
+    struct timespec *timeout)
+{
+    int t = uv_backend_timeout(data);
+
+    SOL_DBG("t = %d", t);
```

Comment:
I can change the function to not do the math if it's going to return false anyway.